### PR TITLE
Add example for `FieldBoundaryConditions` on an `ImmersedBoundaryGrid`

### LINKED ISSUE #2631: Add example for `FieldBoundaryConditions` on an `ImmersedBoundaryGrid`
The typical example for this should be a linear bottom drag, e.g., for `u` a condition `- μ u`.

Additionally, the `FieldBoundaryConditions` [docstring](https://github.com/CliMA/Oceananigans.jl/blob/42b2130b0945956632624a17b4d2d28631e1eee4/src/BoundaryConditions/field_boundary_conditions.jl#L103-L128) needs some update to include the `immersed` kwarg.

cc @txs156, @glwagner

### PR DESCRIPTION
**Pull request by navidcy**
Closes #2631

### CODE REVIEW

**Review by francispoulin — APPROVED** [review]

**Comment by simone-silvestri** [comment]
I think the immersed boundary condition should be an `ImmersedBoundaryCondition`

```
@inline linear_drag(i, j, k, grid, clock, fields) = - 0.2 * fields.u[i, j, k]

drag_u = FluxBoundaryCondition(linear_drag, discrete_form=true)

u_immersed_bc = ImmersedBoundaryCondition(bottom = drag_u)

u_bcs = FieldBoundaryConditions(bottom = drag_u, immersed = u_immersed_bc)
```

**Comment by navidcy** [comment]
So shall I merge then? It's correct?

**Comment by navidcy** [comment]
@simone-silvestri can you re-review?

File: docs/src/model_setup/boundary_conditions.md [hunk 1]
```diff
@@ -488,3 +488,31 @@ Oceananigans.FieldBoundaryConditions, with boundary conditions
 └── immersed: ImmersedBoundaryCondition with west=Nothing, east=Nothing, south=Nothing, north=Nothing, bottom=Value, top=Nothing
 ```
 
+A boundary condition that depends on the fields may be prescribed using the `immersed` keyword argument in [`FieldBoundaryConditions`](@ref), e.g.,
+
+```jldoctest
+julia> @inline linear_drag(x, y, z, t, u) = - 0.2 * u
```
Comment (simone-silvestri) [inline docs/src/model_setup/boundary_conditions.md:494]:
```suggestion
julia> @inline linear_drag(x, y, t, u) = - 0.2 * u
```

File: docs/src/model_setup/boundary_conditions.md [hunk 2]
```diff
@@ -488,3 +488,31 @@ Oceananigans.FieldBoundaryConditions, with boundary conditions
...
+linear_drag (generic function with 1 method)
+
+julia> drag_u = FluxBoundaryCondition(linear_drag, field_dependencies=:u)
+FluxBoundaryCondition: ContinuousBoundaryFunction linear_drag at (Nothing, Nothing, Nothing)
+
+julia> u_immersed_bc = ImmersedBoundaryCondition(bottom = drag_u)
```
Comment (simone-silvestri) [inline docs/src/model_setup/boundary_conditions.md:500]:
```suggestion
julia> @inline linear_drag(x, y, t, u) = - 0.2 * u
linear_drag (generic function with 1 method)

julia> @inline immersed_linear_drag(x, y, z, t, u) = - 0.2 * u
immersed_linear_drag (generic function with 1 method)

julia> drag_u = FluxBoundaryCondition(linear_drag, field_dependencies=:u)
FluxBoundaryCondition: ContinuousBoundaryFunction linear_drag at (Nothing, Nothing, Nothing)

julia> immersed_drag_u = FluxBoundaryCondition(immersed_linear_drag, field_dependencies=:u)
FluxBoundaryCondition: ContinuousBoundaryFunction immersed_linear_drag at (Nothing, Nothing, Nothing)

julia> u_immersed_bc = ImmersedBoundaryCondition(bottom = immersed_drag_u)
ImmersedBoundaryCondition:
├── west: Nothing
├── east: Nothing
├── south: Nothing
├── north: Nothing
├── bottom: FluxBoundaryCondition: ContinuousBoundaryFunction immersed_linear_drag at (Nothing, Nothing, Nothing)
└── top: Nothing

julia> u_bcs = FieldBoundaryConditions(bottom = drag_u, immersed = u_immersed_bc)
Oceananigans.FieldBoundaryConditions, with boundary conditions
├── west: DefaultBoundaryCondition (FluxBoundaryCondition: Nothing)
├── east: DefaultBoundaryCondition (FluxBoundaryCondition: Nothing)
├── south: DefaultBoundaryCondition (FluxBoundaryCondition: Nothing)
├── north: DefaultBoundaryCondition (FluxBoundaryCondition: Nothing)
├── bottom: FluxBoundaryCondition: ContinuousBoundaryFunction linear_drag at (Nothing, Nothing, Nothing)
├── top: DefaultBoundaryCondition (FluxBoundaryCondition: Nothing)
└── immersed: ImmersedBoundaryCondition with west=Nothing, east=Nothing, south=Nothing, north=Nothing, bottom=Flux, top=Nothing
```

Comment (simone-silvestri) [reply docs/src/model_setup/boundary_conditions.md:500]: sorry I messed up again the multiline

Comment (navidcy) [reply docs/src/model_setup/boundary_conditions.md]: No worries! Done by [3c8c748](https://github.com/CliMA/Oceananigans.jl/pull/2632/commits/3c8c7486ecbc8d6cb221e1aea41c86f6820f0428)

File: docs/src/model_setup/boundary_conditions.md [hunk 3]
```diff
@@ -488,3 +488,49 @@ Oceananigans.FieldBoundaryConditions, with boundary conditions
...
+linear_drag (generic function with 1 method)
+
+julia> drag_u = FluxBoundaryCondition(linear_drag, field_dependencies=:u)
+FluxBoundaryCondition: ContinuousBoundaryFunction linear_drag at (Nothing, Nothing, Nothing)
+```
+
+Then let's create the immersed boundary condition:
+
+```jldoctest immersed_bc
+julia> @inline immersed_linear_drag(x, y, z, t, u) = - 0.2 * u
```
Comment (tomchor) [inline docs/src/model_setup/boundary_conditions.md:508]: I think something that may not be intuitive and that might be worth calling attention to is that the function for the bottom only depends on `(x, y, t, u)`, but the one for the immersed boundary also depends on `z`.

Comment (navidcy) [reply docs/src/model_setup/boundary_conditions.md:508]: Good point!

**Comment by glwagner** [comment]
> I think the immersed boundary condition should be an `ImmersedBoundaryCondition`
> 
> ```
> @inline linear_drag(i, j, k, grid, clock, fields) = - 0.2 * fields.u[i, j, k]
> 
> drag_u = FluxBoundaryCondition(linear_drag, discrete_form=true)
> u_immersed_bc = ImmersedBoundaryCondition(bottom = drag_u)
> u_bcs = FieldBoundaryConditions(bottom = drag_u, immersed = u_immersed_bc)
> ```
> 
> I do not think we can use the continuous form because `ImmersedBoundaryCondition` should behave just as another `FieldBoundaryCondition` which is 3D instead of 2D and I do not see a `getbc` defined for a `3DBoundaryFunction` (they are only defined for `XBoundaryFunction`, `YBoundaryFunction` and `ZBoundaryFunction`)
> 
> we can probably add here the `getbc` method for a _continuous_ immersed boundary condition

This is a specific choice that depends on the boundary condition one is trying to implement. For "simple" boundary conditions, we don't need to invoke `ImmersedBoundaryCondition` because

https://github.com/CliMA/Oceananigans.jl/blob/95206ae73e10460beebb2ce9d5fc9945523e5724/src/ImmersedBoundaries/immersed_boundary_condition.jl#L242-L245

For boundary conditions that impose _only bottom drag_ but _not_ side drag, then we do need to use `ImmersedBoundaryCondition` as indicated.

File: docs/src/model_setup/boundary_conditions.md (hunk 3 above)
Comment (glwagner) [reply docs/src/model_setup/boundary_conditions.md:508]: I'd just point out that neither is "intuitive", they simply are what they are. It'd good to point out this difference though.

**Comment by navidcy** [comment]
> > I think the immersed boundary condition should be an `ImmersedBoundaryCondition`
> > ```
> > @inline linear_drag(i, j, k, grid, clock, fields) = - 0.2 * fields.u[i, j, k]
> > 
> > drag_u = FluxBoundaryCondition(linear_drag, discrete_form=true)
> > u_immersed_bc = ImmersedBoundaryCondition(bottom = drag_u)
> > u_bcs = FieldBoundaryConditions(bottom = drag_u, immersed = u_immersed_bc)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I do not think we can use the continuous form because `ImmersedBoundaryCondition` should behave just as another `FieldBoundaryCondition` which is 3D instead of 2D and I do not see a `getbc` defined for a `3DBoundaryFunction` (they are only defined for `XBoundaryFunction`, `YBoundaryFunction` and `ZBoundaryFunction`)
> > we can probably add here the `getbc` method for a _continuous_ immersed boundary condition
> 
> This is a specific choice that depends on the boundary condition one is trying to implement. For "simple" boundary conditions, we don't need to invoke `ImmersedBoundaryCondition` because
> 
> https://github.com/CliMA/Oceananigans.jl/blob/95206ae73e10460beebb2ce9d5fc9945523e5724/src/ImmersedBoundaries/immersed_boundary_condition.jl#L242-L245
> 
> For boundary conditions that impose _only bottom drag_ but _not_ side drag, then we do need to use `ImmersedBoundaryCondition` as indicated.

Shall I leave it like this? Bottom drag is usually applied at the bottom (not the sides) so the example is relevant and useful to many applications. Is it too elaborate though?

File: docs/src/model_setup/boundary_conditions.md (hunk 3 above)
Comment (navidcy) [reply docs/src/model_setup/boundary_conditions.md:508]: resolved by [d00e735](https://github.com/CliMA/Oceananigans.jl/pull/2632/commits/d00e73581a91e0479411ee0c818ce6f201143a55)

**Comment by glwagner** [comment]
> > > I think the immersed boundary condition should be an `ImmersedBoundaryCondition`
> > > ```
> > > @inline linear_drag(i, j, k, grid, clock, fields) = - 0.2 * fields.u[i, j, k]
> > > 
> > > drag_u = FluxBoundaryCondition(linear_drag, discrete_form=true)
> > > u_immersed_bc = ImmersedBoundaryCondition(bottom = drag_u)
> > > u_bcs = FieldBoundaryConditions(bottom = drag_u, immersed = u_immersed_bc)
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > I do not think we can use the continuous form because `ImmersedBoundaryCondition` should behave just as another `FieldBoundaryCondition` which is 3D instead of 2D and I do not see a `getbc` defined for a `3DBoundaryFunction` (they are only defined for `XBoundaryFunction`, `YBoundaryFunction` and `ZBoundaryFunction`)
> > > we can probably add here the `getbc` method for a _continuous_ immersed boundary condition
> > 
> > 
> > This is a specific choice that depends on the boundary condition one is trying to implement. For "simple" boundary conditions, we don't need to invoke `ImmersedBoundaryCondition` because
> > https://github.com/CliMA/Oceananigans.jl/blob/95206ae73e10460beebb2ce9d5fc9945523e5724/src/ImmersedBoundaries/immersed_boundary_condition.jl#L242-L245
> > 
> > For boundary conditions that impose _only bottom drag_ but _not_ side drag, then we do need to use `ImmersedBoundaryCondition` as indicated.
> 
> Shall I leave it like this? Bottom drag is usually applied at the bottom (not the sides) so the example is relevant and useful to many applications. Is it too elaborate though?

I think if its a common application, it's important to demonstrate.

File: docs/src/model_setup/boundary_conditions.md [hunk 4]
```diff
@@ -488,3 +488,54 @@ Oceananigans.FieldBoundaryConditions, with boundary conditions
...
+├── north: DefaultBoundaryCondition (FluxBoundaryCondition: Nothing)
+├── bottom: FluxBoundaryCondition: ContinuousBoundaryFunction linear_drag at (Nothing, Nothing, Nothing)
+├── top: DefaultBoundaryCondition (FluxBoundaryCondition: Nothing)
+└── immersed: ImmersedBoundaryCondition with west=Nothing, east=Nothing, south=Nothing, north=Nothing, bottom=Flux, top=Nothing
+```
+
+!!! warn"Positional argument requirements"
+    Note the difference between the arguments required for the function within the `bottom` boundary
+    condition versus the arguments for the function within the `immersed` boundary condition. E.g.,
+    `x, y, t` in linear_drag() versus `x, y, z, t` in `immersed_linear_drag()`.
```
Comment (tomchor) [inline docs/src/model_setup/boundary_conditions.md:541]:
```suggestion
    `x, y, t` in `linear_drag()` versus `x, y, z, t` in `immersed_linear_drag()`.
```

File: docs/src/model_setup/boundary_conditions.md [hunk 5]
```diff
@@ -488,3 +488,54 @@ Oceananigans.FieldBoundaryConditions, with boundary conditions
...
+FluxBoundaryCondition: ContinuousBoundaryFunction linear_drag at (Nothing, Nothing, Nothing)
+```
+
+Then let's create the immersed boundary condition:
+
+```jldoctest immersed_bc
+julia> @inline immersed_linear_drag(x, y, z, t, u) = - 0.2 * u
+immersed_linear_drag (generic function with 1 method)
+
+julia> immersed_drag_u = FluxBoundaryCondition(immersed_linear_drag, field_dependencies=:u)
```
Comment (tomchor) [inline docs/src/model_setup/boundary_conditions.md:511]:
```suggestion
julia> @inline linear_drag(x, y, z, t, u) = - 0.2 * u
linear_drag (generic function with 1 method)

julia> immersed_drag_u = FluxBoundaryCondition(linear_drag, field_dependencies=:u)
```

Just a suggestion, but using the same name `linear_drag` for both functions makes the code cleaner and illustrates that we can rely on multiple dispatch to use the correct method

Comment (glwagner) [reply docs/src/model_setup/boundary_conditions.md:511]: One downside is that using the same function name might be confusing for people who aren't well-versed in multiple dispatch

**Comment by navidcy** [comment]
Actually, @glwagner, what exactly do you mean by "applied at the bottom VS applied at the sides"? For example, in a situation like the one below, the cell marked with black the bottom drag will have to be applied on the cell's southern side. Does this fall under what you mean by "apply also at the sides"? If so, then that's what we want!

Note Jul 13, 2022

**Comment by glwagner** [comment]
> Actually, @glwagner, what exactly do you mean by "applied at the bottom VS applied at the sides"? For example, in a situation like the one below, the cell marked with black the bottom drag will have to be applied on the cell's southern side. Does this fall under what you mean by "apply also at the sides"? If so, then that's what we want!
> 
> Note Jul 13, 2022

I don't completely understand the question but I think @simone-silvestri's statements are correct. An `ImmersedBoundaryCondition` has fields that represent fluxes across _all_ six sides of the cell indepedently. For every cell, the specified immersed flux is applied if that _particular_ face is diagnosed as a boundary between an immersed solid, and fluid. Therefore one can apply fluxes at the "sides" (ie south, north, east, west) independently from fluxes across the "bottom".

**Comment by navidcy** [comment]
Gotcha

File: docs/src/model_setup/boundary_conditions.md (hunk 5 above)
Comment (navidcy) [reply docs/src/model_setup/boundary_conditions.md:511]: I left it w different names.

**Pull request merged**